File: hw/hw11/hw11.py
#!/usr/bin/env python
# coding: utf-8

# In[7]:


from queue import Queue


# In[8]:


def floodFill(image, sr, sc, color):
    #print(f" img: {image[sr][sc]} col: {color}")
    if image[sr][sc] == color:
        print("unchanged")
        return image
    
    originalColor = image[sr][sc]
    
    queue = Queue()
    queue.put((sr, sc))
    
    while not queue.empty():
        currentPixel = queue.get()

        # loc pixel
        x, y = currentPixel

        # ignore out of bounds x,y, or if different from original color
        if x < 0 or x >= len(image) or y < 0 or y >= len(image[0]) or image[x][y] != originalColor:
            continue
                
        # sets new color
        image[x][y] = color
        
        # adds adjacent pixels to queue
        queue.put((x + 1, y))
        queue.put((x - 1, y))
        queue.put((x, y + 1))
        queue.put((x, y - 1)) 
        
    
    return image


# In[9]:


def printMatrix(image):
    for row in image:
        print(row)


# ## Test Case

# In[10]:


image = [[1,1,1],
        [1,1,0],
        [1,0,1]]

# sr,sc = starting position, at (1,1) this indicates it'll start in the middle of the matrix and expand.
# Then pixels will be colored as long as they're adjacent to another pixel.

sr,sc,color=1,1,2
print("Original Matrix")
printMatrix(image)
flooded = floodFill(image,sr,sc,color)

print("\nFlooded Matrix")
printMatrix(flooded)


# In[11]:


image = [[0,1,2],
        [0,4,5],
        [0,7,8]]
# sr,sc = starting position, at (1,1) this indicates it'll start in the middle of the matrix and expand.
# Then pixels will be colored as long as they're adjacent to another pixel.

sr,sc,color=0,0,1
print("Original Matrix")
printMatrix(image)
flooded = floodFill(image,sr,sc,color)

print("\nFlooded Matrix")
printMatrix(flooded)


# In[12]:


image = [[0,0,0],
         [0,0,0]]
sr,sc,color = 0,0,1

print("Original Matrix")
printMatrix(image)
flooded = floodFill(image,sr,sc,color)

print("\nFlooded Matrix")
printMatrix(flooded)


# In[ ]:


get_ipython().system('jupyter nbconvert --to script hw11.ipynb')

File: hw/hw11/hw11.java


import java.util.LinkedList;
import java.util.Queue;

public class hw11 {

    public static int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if (image[sr][sc] == newColor) {
            System.out.println("unchanged");
            return image;
        }

        int originalColor = image[sr][sc];
        int[][] directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{sr, sc});

        while (!queue.isEmpty()) {
            int[] currentPixel = queue.poll();
            int x = currentPixel[0];
            int y = currentPixel[1];

            if (x < 0 || x >= image.length || y < 0 || y >= image[0].length || image[x][y] != originalColor) {
                continue;
            }

            image[x][y] = newColor;

            for (int[] dir : directions) {
                queue.add(new int[]{x + dir[0], y + dir[1]});
            }
        }

        return image;
    }
    public static void printFloodedMatrix(String label, int[][] originalImage, int sr, int sc, int color) {
        System.out.println(label + ":\nOriginal Matrix:");
        printMatrix(originalImage);
        int[][] floodedImage = floodFill(originalImage, sr, sc, color);
        System.out.println("\nFlooded Matrix:");
        printMatrix(floodedImage);
        System.out.println("");
    }
    
    public static void printMatrix(int[][] image) {
        for (int[] row : image) {
            for (int pixel : row) {
                System.out.print(pixel + " ");
            }
            System.out.println();
        }
    }
    public static void main(String[] args) {
        int[][] image1 = {{1, 1, 1}, {1, 1, 0}, {1, 0, 1}};
        int sr1 = 1, sc1 = 1, color1 = 2;
    
        printFloodedMatrix("Case 1", image1, sr1, sc1, color1);
    
        int[][] image2 = {{0, 1, 2}, {0, 4, 5}, {0, 7, 8}};
        int sr2 = 0, sc2 = 0, color2 = 1;
    
        printFloodedMatrix("Case 2", image2, sr2, sc2, color2);
    
        int[][] image3 = {{0, 0, 0}, {0, 0, 0}};
        int sr3 = 0, sc3 = 0, color3 = 1;
    
        printFloodedMatrix("Case 3", image3, sr3, sc3, color3);
    }
    
    
    
}
